File: solutions/java/0126.java
class Solution {
  public List<List<String>> findLadders(String beginWord, String endWord, List<String> wordList) {
    List<List<String>> ans = new ArrayList<>();
    Set<String> set = new HashSet<>(wordList);

    if (!set.contains(endWord))
      return ans;

    Set<String> set1 = new HashSet<>(Arrays.asList(beginWord));
    Map<String, List<String>> map = new HashMap<>();
    boolean isFound = false;

    while (!set1.isEmpty() && !isFound) {
      for (String word : set1)
        set.remove(word);
      Set<String> tempSet = new HashSet<>();
      for (String parent : set1) {
        char[] word = parent.toCharArray();
        for (int i = 0; i < word.length; ++i) {
          char c = word[i];
          for (char j = 'a'; j <= 'z'; ++j) {
            word[i] = j;
            String newWord = String.valueOf(word);
            if (newWord.equals(endWord)) {
              if (!map.containsKey(parent))
                map.put(parent, new ArrayList<>());
              map.get(parent).add(newWord);
              isFound = true;
            } else if (set.contains(newWord) && !isFound) {
              tempSet.add(newWord);
              if (!map.containsKey(parent))
                map.put(parent, new ArrayList<>());
              map.get(parent).add(newWord);
            }
          }
          word[i] = c;
        }
      }
      set1 = tempSet;
    }

    if (isFound) {
      List<String> path = new ArrayList<>(Arrays.asList(beginWord));
      dfs(map, beginWord, endWord, path, ans);
    }

    return ans;
  }

  private void dfs(Map<String, List<String>> map, String word, String endWord, List<String> path,
      List<List<String>> ans) {
    if (word.equals(endWord)) {
      ans.add(new ArrayList<>(path));
      return;
    }
    if (!map.containsKey(word))
      return;

    for (String child : map.get(word)) {
      path.add(child);
      dfs(map, child, endWord, path, ans);
      path.remove(path.size() - 1);
    }
  }
}

File: solutions/cpp/0126.cpp
class Solution {
 public:
  vector<vector<string>> findLadders(string beginWord, string endWord,
                                     vector<string>& wordList) {
    unordered_set<string> wordListSet(begin(wordList), end(wordList));
    if (!wordListSet.count(endWord)) return {}; // prune (still AC if this line is removed)

    vector<vector<string>> ans;
    unordered_map<string, vector<string>> parentToWords;
    unordered_set<string> currentLevelWords{beginWord};
    bool isFound = false;

    while (!currentLevelWords.empty() && !isFound) {
      // remove words in current level
      for (const string& word : currentLevelWords)
        wordListSet.erase(word);
      unordered_set<string> nextLevelWords;
      // `parent` will be used as a key in `parentToWords`
      for (const string& parent : currentLevelWords) {
        string nextWord = parent;
        // enumerate next level words
        for (int i = 0; i < nextWord.length(); ++i) {
          const char cache = nextWord[i];
          for (char c = 'a'; c <= 'z'; ++c) {
            nextWord[i] = c;
            if (wordListSet.count(nextWord)) {
              if (nextWord == endWord)
                isFound = true;
              nextLevelWords.insert(nextWord);
              parentToWords[parent].push_back(nextWord);
            }
          }
          nextWord[i] = cache;
        }
      }
      if (isFound) break;
      swap(currentLevelWords, nextLevelWords);
    }

    // construct the ans by `parentToWords`
    function<void(const string&, vector<string>&)> dfs =
        [&](const string& word, vector<string>& path) {
          if (word == endWord) {
            ans.push_back(path);
            return;
          }
          if (!parentToWords.count(word)) return;

          for (const string& child : parentToWords[word]) {
            path.push_back(child);
            dfs(child, path);
            path.pop_back();
          }
        };

    if (isFound) {
      vector<string> path{beginWord};
      dfs(beginWord, path);
    }

    return ans;
  }
};